qiskit_aqua/algorithms/adaptive/qsvm/cost_helpers.py

```diff
@@ -82,6 +81,50 @@ def cost_estimate_sigmoid(shots, probs, gt_labels):
     Returns:
         float: averaged sigmoid cross entropy loss between estimated probs and gt_labels
     """
+    """
+    The prior version of this function errors out when the probabilities in other_probs have 
+    repeated values for most entries. For example, the bug can be simulated when 
+    the original version of cost_estimate_sigmoid is called with the below values for the 
+    function parameters.
+    probs=np.array([[0.14, 0.25, 0.45], [0.15, 0.25, 0.45], [0.14, 0.25, 0.14]])
+    gt_labels = np.array([0, 2, 1])
+
+    Specifically, the second line below errors out with a dimensionality error when
+    the setdiff1d call results in a single element array for other_probs,
+    and so other_probs[1] causes an error.
+    The logic for cross-entropy loss is not correct in this scenario. The calculation for 
+    cross-entropy loss has been fixed as part of this fork.
+    other_probs = np.setdiff1d(probs[other_index], rest_p)
+        numerator = np.sqrt(shots) * ((1. + np.abs(other_probs[0] - other_probs[1])) /
+                                      number_of_classes - rest_p)
+    """
+    epsilon = 1e-15
+    probs = np.clip(probs, epsilon, 1 - epsilon)
+    number_of_classes = probs.shape[1]
+
+    if number_of_classes > 1:
+        #if len(np.unique(gt_labels)) > number_of_classes:
+        #    raise ValueError('Too many classes compared to pred.')
+        y_true = np.eye(number_of_classes)[np.array(gt_labels).reshape(-1)]
+    else:
+        #if len(np.unique(gt_labels)) > 2:
+        #    raise ValueError('Too many classes compared to pred.')
+        y_true = gt_labels
+        loss = -1 * (y_true * np.log(probs) + (1 - y_true) * np.log(1 - probs)).sum(axis=1)
+        return loss
+
+    if y_true.shape[1] != probs.shape[1]:
+        raise ValueError('Pred and y_true have different number of classes.')
+    if y_true.shape[0] != probs.shape[0]:
+        raise ValueError('Pred and y_true have different number of samples.')
+
+    probs /= probs.sum(axis=1)[:, np.newaxis]
+    loss = -1 * (y_true * np.log(probs)).sum(axis=1)
+    loss = np.sqrt(shots) * np.average(loss)
```

Comment:
any explanation for why we need to multiply the standard cross_entropy with np.sqrt(shots)? 
according to the tests, it works well. but i am curious why it helps.

Reply:
The original author of the code may have more insight into why the sqrt(shots) multiplier is applied since this was included in the original loss function code. Below is my understanding of why it may be relevant to apply this multiplier.
The training loss optimization itself does not really get affected or changed by the extra multiplication of sqrt(shots) which is basically a constant multiplier, so the net effect on the training and cost optimization and hence accuracy does not change if we include the constant multiplier or not.
However, the below is why I think the extra constant multiplier is applied to the loss.
When the run_algorithm call is made, these are the steps that happen in sequence:
1. Set up and initialisation of the quantum algorithm, quantum backend setup etc.
2. Train is executed with the training data set (no quantum circuit involved at this point). Training causes the cost optimizer to minimize the training loss to convergence
3. Test is executed which calls predict on the test data set. This is where the quantum circuit gets involved and the number of measurements taken is = shots before the predicted probabilities are extracted based on the counts from the quantum circuit.
So though the training loss optimization is unaffected by the constant shots multiplier, for a more accurate representation of the actual value of the training loss itself, it may make sense to include the multiplier of shots given the test predictions are taken n=shots times before the test predictions are generated.